src/TransitionSystems/quotient.rs

```diff
@@ -1,37 +1,497 @@
+            T.get_input_actions(),
+            T.get_output_actions()
+        );
+
+        let ts = Box::new(Quotient {
+            T,
+            S,
+            inputs,
+            outputs,
+            universal_location,
+            inconsistent_location,
+            decls,
+            new_clock_index,
+            new_input_name,
+            dim,
+        });
+        Ok(ts)
+
+        //Ok(pruning::prune_system(ts, dim))
     }
 }
 
-impl TransitionSystem<'static> for Quotient {
-    default_composition!();
-    fn next_transitions<'b>(
-        &'b self,
-        location: &LocationTuple<'b>,
-        action: &str,
-        sync_type: &SyncType,
-        index: &mut usize,
-    ) -> Vec<Transition<'b>> {
-        panic!("Not implemented");
+impl TransitionSystem for Quotient {
+    fn next_transitions(&self, location: &LocationTuple, action: &str) -> Vec<Transition> {
+        //println!("Action: {}", action);
+        assert!(self.actions_contain(action));
+        let is_input = self.inputs_contain(action);
+
+        let mut transitions = vec![];
+        let mut reset_all = vec![];
+        for clock in self
+            .get_decls()
+            .iter()
+            .flat_map(|d| d.get_clocks().values())
+            .copied()
+        {
+            reset_all.push(CompiledUpdate {
+                clock_index: clock,
+                value: 0,
+            });
+        }
+
+        //Rules [universal] and [inconsistent]
+
+        if location.is_inconsistent() {
+            //Rule 10
+            if is_input {
+                let mut transition = Transition::new(location, self.dim);
+                transition
+                    .guard_zone
+                    .add_eq_const_constraint(self.new_clock_index, 0);
+                transitions.push(transition);
+            }
+            return transitions;
+        } else if location.is_universal() {
+            // Rule 9
+            let mut transition = Transition::new(location, self.dim);
+            /* This should be okay */
+            transition.updates = reset_all.clone();
+            transitions.push(transition);
+            return transitions;
+        }
+
+        // As it is not universal or inconsistent it must be a quotient loc
+        let loc_t = location.get_left();
+        let loc_s = location.get_right();
+        let t = self.T.next_transitions_if_available(loc_t, action);
+        let s = self.S.next_transitions_if_available(loc_s, action);
+
+        let inconsistent_location =
+            LocationTuple::simple(&self.inconsistent_location, &self.decls, self.dim);
+        let universal_location =
+            LocationTuple::simple(&self.universal_location, &self.decls, self.dim);
+
+        //Rule 1
+        if self.S.actions_contain(action) && self.T.actions_contain(action) {
+            for t_transition in &t {
+                for s_transition in &s {
+                    // ϕ_T ∧ Inv(l2_t)[r |-> 0] ∧ Inv(l1_t)
```

Comment:
Cool idea using the same characters as in the paper, should we mention that phi means guard in a comment?

Reply:
I added a small explanation when it is used